### ISSUE TITLE
[EXPL] [Opt] - add LR scheduler support

### ISSUE THREAD

**Issue by Mars-Martiny** [Mars-Martiny/CITS5206_group1#69]

**Comment by aashutosh-chapagain** [comment]
Also touches issue #100 but with different LR scheduler support.

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>run_name</th>
      <th>tag</th>
      <th>best_epoch</th>
      <th>best_metric_name</th>
      <th>best_metric_value</th>
      <th>learning_rate</th>
      <th>dropout</th>
      <th>weight_decay</th>
      <th>max_length</th>
      <th>pooling</th>
      <th>use_class_weights</th>
      <th>scheduler</th>
      <th>scheduler_config</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>5</th>
      <td>safetybert_sched_06_do025_plateau_lr2e-05_do0....</td>
      <td>do025_plateau</td>
      <td>3</td>
      <td>f1_macro</td>
      <td>0.662599</td>
      <td>0.000020</td>
      <td>0.25</td>
      <td>0.01</td>
      <td>160</td>
      <td>cls</td>
      <td>False</td>
      <td>ReduceLROnPlateau</td>
      <td>{'name': 'ReduceLROnPlateau', 'monitor': 'f1_m...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>safetybert_sched_03_lr3e5_plateau_lr3e-05_do0....</td>
      <td>lr3e5_plateau</td>
      <td>4</td>
      <td>f1_macro</td>
      <td>0.650652</td>
      <td>0.000030</td>
      <td>0.30</td>
      <td>0.01</td>
      <td>160</td>
      <td>cls</td>
      <td>False</td>
      <td>ReduceLROnPlateau</td>
      <td>{'name': 'ReduceLROnPlateau', 'monitor': 'f1_m...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>safetybert_sched_02_cosine_lr2e-05_do0.3_wd0.0...</td>
      <td>cosine</td>
      <td>4</td>
      <td>f1_macro</td>
      <td>0.645766</td>
      <td>0.000020</td>
      <td>0.30</td>
      <td>0.01</td>
      <td>160</td>
      <td>cls</td>
      <td>False</td>
      <td>CosineAnnealingLR</td>
      <td>{'name': 'CosineAnnealingLR', 'T_max': 6, 'eta...</td>
    </tr>
    <tr>
      <th>6</th>
      <td>safetybert_sched_07_ml192_plateau_lr2e-05_do0....</td>
      <td>ml192_plateau</td>
      <td>5</td>
      <td>f1_macro</td>
      <td>0.640361</td>
      <td>0.000020</td>
      <td>0.30</td>
      <td>0.01</td>
      <td>192</td>
      <td>cls</td>
      <td>False</td>
      <td>ReduceLROnPlateau</td>
      <td>{'name': 'ReduceLROnPlateau', 'monitor': 'f1_m...</td>
    </tr>
    <tr>
      <th>4</th>
      <td>safetybert_sched_05_cw_plateau_lr2e-05_do0.3_w...</td>
      <td>cw_plateau</td>
      <td>2</td>
      <td>f1_macro</td>
      <td>0.638883</td>
      <td>0.000020</td>
      <td>0.30</td>
      <td>0.01</td>
      <td>160</td>
      <td>cls</td>
      <td>True</td>
      <td>ReduceLROnPlateau</td>
      <td>{'name': 'ReduceLROnPlateau', 'monitor': 'f1_m...</td>
    </tr>
    <tr>
      <th>0</th>
      <td>safetybert_sched_01_plateau_f1_lr2e-05_do0.3_w...</td>
      <td>plateau_f1</td>
      <td>5</td>
      <td>f1_macro</td>
      <td>0.619671</td>
      <td>0.000020</td>
      <td>0.30</td>
      <td>0.01</td>
      <td>160</td>
      <td>cls</td>
      <td>False</td>
      <td>ReduceLROnPlateau</td>
      <td>{'name': 'ReduceLROnPlateau', 'monitor': 'f1_m...</td>
    </tr>
    <tr>
      <th>3</th>
      <td>safetybert_sched_04_lr15e6_cosine_lr1.5e-05_do...</td>
      <td>lr15e6_cosine</td>
      <td>3</td>
      <td>f1_macro</td>
      <td>0.609904</td>
      <td>0.000015</td>
      <td>0.30</td>
      <td>0.01</td>
      <td>160</td>
      <td>cls</td>
      <td>False</td>
      <td>CosineAnnealingLR</td>
      <td>{'name': 'CosineAnnealingLR', 'T_max': 6, 'eta...</td>
    </tr>
  </tbody>
</table>
</div>


best configuration so far for the SafetyBERT with 0.66 f1 score

```
{
    "learning_rate": 2e-5,
    "dropout": 0.25,
    "weight_decay": 0.01,
    "max_length": 160,
    "pooling": "cls",
    "use_class_weights": False,
    "scheduler": "ReduceLROnPlateau",
    "monitor": "f1_macro",
}
```

**Comment by aashutosh-chapagain** [comment]
more results on hyperparameter search with LR scheduling:

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>run_name</th>
      <th>tag</th>
      <th>best_epoch</th>
      <th>best_metric_name</th>
      <th>best_metric_value</th>
      <th>learning_rate</th>
      <th>dropout</th>
      <th>weight_decay</th>
      <th>max_length</th>
      <th>pooling</th>
      <th>use_class_weights</th>
      <th>scheduler</th>
      <th>scheduler_config</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2</th>
      <td>safetybert_sched_03_do0275_plateau_lr2e5_lr2e-...</td>
      <td>do0275_plateau_lr2e5</td>
      <td>4</td>
      <td>f1_macro</td>
      <td>0.664149</td>
      <td>0.000020</td>
      <td>0.275</td>
      <td>0.01</td>
      <td>160</td>
      <td>cls</td>
      <td>False</td>
      <td>ReduceLROnPlateau</td>
      <td>{'name': 'ReduceLROnPlateau', 'monitor': 'f1_m...</td>
    </tr>
    <tr>
      <th>0</th>
      <td>safetybert_sched_01_do020_plateau_lr2e5_lr2e-0...</td>
      <td>do020_plateau_lr2e5</td>
      <td>5</td>
      <td>f1_macro</td>
      <td>0.655214</td>
      <td>0.000020</td>
      <td>0.200</td>
      <td>0.01</td>
      <td>160</td>
      <td>cls</td>
      <td>False</td>
      <td>ReduceLROnPlateau</td>
      <td>{'name': 'ReduceLROnPlateau', 'monitor': 'f1_m...</td>
    </tr>
    <tr>
      <th>4</th>
      <td>safetybert_sched_05_do025_plateau_lr3e5_lr3e-0...</td>
      <td>do025_plateau_lr3e5</td>
      <td>4</td>
      <td>f1_macro</td>
      <td>0.647603</td>
      <td>0.000030</td>
      <td>0.250</td>
      <td>0.01</td>
      <td>160</td>
      <td>cls</td>
      <td>False</td>
      <td>ReduceLROnPlateau</td>
      <td>{'name': 'ReduceLROnPlateau', 'monitor': 'f1_m...</td>
    </tr>
    <tr>
      <th>3</th>
      <td>safetybert_sched_04_do025_plateau_lr25e6_lr2.5...</td>
      <td>do025_plateau_lr25e6</td>
      <td>4</td>
      <td>f1_macro</td>
      <td>0.626945</td>
      <td>0.000025</td>
      <td>0.250</td>
      <td>0.01</td>
      <td>160</td>
      <td>cls</td>
      <td>False</td>
      <td>ReduceLROnPlateau</td>
      <td>{'name': 'ReduceLROnPlateau', 'monitor': 'f1_m...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>safetybert_sched_02_do025_plateau_lr2e5_repeat...</td>
      <td>do025_plateau_lr2e5_repeat</td>
      <td>5</td>
      <td>f1_macro</td>
      <td>0.614507</td>
      <td>0.000020</td>
      <td>0.250</td>
      <td>0.01</td>
      <td>160</td>
      <td>cls</td>
      <td>False</td>
      <td>ReduceLROnPlateau</td>
      <td>{'name': 'ReduceLROnPlateau', 'monitor': 'f1_m...</td>
    </tr>
  </tbody>
</table>
</div>

**Comment by aashutosh-chapagain** [comment]
```
BEST_SAFETYBERT_CONFIG = {
    "fine_tune": True,
    "pooling": "cls",
    "learning_rate": 2e-5,
    "dropout": 0.275,
    "weight_decay": 0.01,
    "max_length": 160,
    "batch_size": 8,
    "epochs": 6,
    "use_class_weights": False,
    "scheduler_config": {
        "name": "ReduceLROnPlateau",
        "monitor": "f1_macro",
        "mode": "max",
        "factor": 0.5,
        "patience": 1,
        "min_lr": 1e-7,
        "step_per_batch": False,
    },
}
```
